File: 1250-longest-common-subsequence/longest-common-subsequence.js
/**
 * @param {string} text1
 * @param {string} text2
 * @return {number}
 */
var longestCommonSubsequence = function (text1, text2) {
    // return lcsRecursive(text1, text2, text1.length - 1, text2.length - 1);
    // const dp = new Array(text1.length).fill().map(() => new Array(text2.length).fill(-1));
    // return lcsMemo(text1, text2, text1.length - 1, text2.length - 1, dp);
    // return lcsTabulate(text1, text2);
    return lcsSpaceOptimized(text1, text2);
};

function lcsSpaceOptimized(text1, text2) {
    const textLen1 = text1.length, textLen2 = text2.length;
    let prev = new Array(textLen2 + 1), curr = new Array(textLen2 + 1).fill(0);
    for (let j = 0; j <= textLen2; j++) {
        prev[j] = 0;
    }

    for (let i = 1; i <= textLen1; i++) {
        for (let j = 1; j <= textLen2; j++) {
            if (text1[i - 1] === text2[j - 1]) {
                // both match
                curr[j] = 1 + prev[j - 1];
            } else {
                // doesn't match
                curr[j] = Math.max(prev[j], curr[j - 1]);
            }
        }
        prev = [...curr]; ///// KEEP IN MIND TO DO LIKE THIS
    }

    return prev[textLen2];
}

function lcsTabulate(text1, text2) {
    const textLen1 = text1.length, textLen2 = text2.length;
    const dp = new Array(textLen1 + 1).fill().map(() => new Array(textLen2 + 1).fill(-1));
    for (let i = 0; i <= textLen1; i++) {
        dp[i][0] = 0;
    }
    for (let j = 0; j <= textLen2; j++) {
        dp[0][j] = 0;
    }

    for (let i = 1; i <= textLen1; i++) {
        for (let j = 1; j <= textLen2; j++) {
            if (text1[i - 1] === text2[j - 1]) {
                // both match
                dp[i][j] = 1 + dp[i - 1][j - 1];
            } else {
                // doesn't match
                dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    return dp[textLen1][textLen2];
}

function lcsMemo(text1, text2, i, j, dp) {
    if (i < 0 || j < 0) {
        return 0;
    }

    if (dp[i][j] !== -1) {
        return dp[i][j];
    }

    if (text1[i] === text2[j]) {
        // both match
        return dp[i][j] = 1 + lcsMemo(text1, text2, i - 1, j - 1, dp);
    } else {
        // doesn't match
        return dp[i][j] = Math.max(lcsMemo(text1, text2, i - 1, j, dp), lcsRecursive(text1, text2, i, j - 1, dp));
    }
}

function lcsRecursive(text1, text2, i, j) {
    if (i < 0 || j < 0) {
        return 0;
    }

    if (text1[i] === text2[j]) {
        // both match
        return 1 + lcsRecursive(text1, text2, i - 1, j - 1);
    } else {
        // doesn't match
        return Math.max(lcsRecursive(text1, text2, i - 1, j), lcsRecursive(text1, text2, i, j - 1));
    }
}

File: 1250-longest-common-subsequence/longest-common-subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int m = text1.size(), n = text2.size();
        // return recursive(text1, text2, m-1, n-1);
        // vector<vector<int>> dp(m, vector<int>(n, -1));
        // return memoize(text1, text2, m-1, n-1, dp);
        return tabulate(text1, text2, m, n);
    }

    int tabulate(string text1, string text2, int m, int n) {
        vector<vector<int>> dp(m+1, vector<int>(n+1, 0));

        for(int i=1; i<=m; i++) {
            for(int j=1; j<=n; j++) {
                if(text1[i-1] == text2[j-1]) {
                    dp[i][j] = dp[i-1][j-1] + 1;
                } else {
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
                }
            }
        }

        return dp[m][n];
    }

    int memoize(string text1, string text2, int i, int j, vector<vector<int>>& dp) {
        int m = text1.size(), n = text2.size();
        if(i == m || j == n) return 0;

        if(i == -1 || j == -1) return 0;

        if(dp[i][j] != -1) return dp[i][j];

        int ans = 0;
        if(text1[i] == text2[j]) {
            ans = memoize(text1, text2, i-1, j-1, dp) + 1;
        } else {
            ans = max(memoize(text1, text2, i-1, j, dp), memoize(text1, text2, i, j-1, dp));
        }

        return dp[i][j] = ans;
    }

    int recursive(string text1, string text2, int i, int j) {
        int m = text1.size(), n = text2.size();
        if(i == m || j == n) return 0;

        if(i == -1 || j == -1) return 0;

        int ans = 0;
        if(text1[i] == text2[j]) {
            ans = recursive(text1, text2, i-1, j-1) + 1;
        } else {
            ans = max(recursive(text1, text2, i-1, j), recursive(text1, text2, i, j-1));
        }

        return ans;
    }
};